core/src/cluster_slots.rs

```diff
@@ -62,16 +62,21 @@ impl ClusterSlots {
                     .zip(std::iter::repeat((epoch_slots.from, stake)))
             })
             .into_group_map();
+        let slot_nodes_stakes: Vec<_> = {
+            let mut cluster_slots = self.cluster_slots.write().unwrap();
```

Comment:
hmm here can we grab the read lock here, clone out the ones that are needed, and then only grab the write lock to insert the missing slots? May avoid grabbing the write lock/blocking in a lot of cases.

Reply:
The older code was doing something like that:
https://github.com/solana-labs/solana/blob/88f22c360/core/src/cluster_slots.rs#L68-L74
and it had a race condition (fixed in https://github.com/solana-labs/solana/pull/15194)

I think that will make the code more convoluted here, and potentially may actually slow it down if it ends up obtaining the write lock anyways. I would suggest to keep this simple for now until we have a benchmark that demonstrates improvement.

Reply:
> The older code was doing something like that:
> https://github.com/solana-labs/solana/blob/88f22c360/core/src/cluster_slots.rs#L68-L74
> and it had a race condition (fixed in #15194)
> 

I see that check needs to check for existence after the write lock is grabbed as well

Then let's add a metric to `RepairStats` in this function call: https://github.com/solana-labs/solana/blob/master/core/src/repair_service.rs#L265, time to grab the read lock in `cluster_slots.compute_weights`. Should give us a benchmark.

Reply:
This is `serve_repair-repair-timing::send-repairs-elapsed`:
https://github.com/solana-labs/solana/blob/d5f235d99/core/src/repair_service.rs#L263-L278
which encompasses the call to `cluster_slots.compute_weights`, and shows that effectively there was no change before and after this change.

This benchmarking was done with the earlier version of this pull request which was extending inner hash-maps inside the write lock (so possibly was even worse).

![btree_index serve_repair-repair-timing send-repairs-elapsed_1](https://user-images.githubusercontent.com/1288998/109083314-9b587080-76fd-11eb-890e-46e1048ee653.png)

Reply:
thanks!